src/psyclone/psyir/symbols/typed_symbol.py

```diff
@@ -0,0 +1,205 @@
+        return self._datatype
+
+    @datatype.setter
+    def datatype(self, value):
+        ''' Setter for the datatype of a TypedSymbol.
+
+        :param value: new value for datatype.
+        :type value: :py:class:`psyclone.psyir.symbols.DataType` or \
+                     :py:class:`psyclone.psyir.symbols.DataTypeSymbol`
+
+        :raises TypeError: if value is not of the correct type.
+        :raises NotImplementedError: if the specified data type is invalid.
+
+        '''
+        # We can't do this import at the toplevel as we get a circular
+        # dependency with the datatypes module.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.symbols import DataType
+        if not isinstance(value, (DataType, DataTypeSymbol)):
+            raise TypeError(
+                "The datatype of a {0} must be specified using either "
+                "a DataType or a DataTypeSymbol but got: '{1}'".format(
+                    type(self).__name__, type(value).__name__))
+        self._datatype = value
+
+    def copy(self):
+        '''Create and return a copy of this object. Any references to the
+        original will not be affected so the copy will not be referred
+        to by any other object.
+
+        :returns: A symbol object with the same properties as this \
+                  symbol object.
+        :rtype: :py:class:`psyclone.psyir.symbols.TypedSymbol`
+
+        '''
+        # The constructors for all Symbol-based classes have 'name' as the
+        # first positional argument.
+        return type(self)(self.name, self.datatype, visibility=self.visibility,
+                          interface=self.interface)
+
+    def copy_properties(self, symbol_in):
+        '''Replace all properties in this object with the properties from
+        symbol_in, apart from the name (which is immutable) and visibility.
+
+        :param symbol_in: the symbol from which the properties are copied.
+        :type symbol_in: :py:class:`psyclone.psyir.symbols.DataSymbol`
+
+        :raises TypeError: if the argument is not the expected type.
+
+        '''
+        if not isinstance(symbol_in, TypedSymbol):
+            raise TypeError("Argument should be of type 'TypedSymbol' but "
+                            "found '{0}'.".format(type(symbol_in).__name__))
+        super(TypedSymbol, self).copy_properties(symbol_in)
+        self._datatype = symbol_in.datatype
+
+    def resolve_deferred(self):
+        ''' If the symbol has a deferred datatype, find where it is defined
+        (i.e. an external container) and obtain the properties of the symbol.
+
+        :returns: this TypedSymbol with its properties updated. This is for \
+                  consistency with the equivalent method in the Symbol \
+                  class which returns a new Symbol object.
+        :rtype: :py:class:`psyclone.psyir.symbols.TypedSymbol`
+
+        '''
+        # This import has to be local to this method to avoid circular
```

Comment:
I am not sure but this may have changed now if the dependency was with the DataTypeSymbol and not the TypedSymbol branch

Reply:
Sadly not. Part of the problem is that Visibility is a property of Symbol. Possibly it should be in Datatypes instead? Besides which, we allow types to be specified as a `DataTypeSymbol` which will probably also cause problems.